rcon/commands.py

```diff
@@ -73,78 +89,119 @@ class ServerCtl:
     set password not implemented on purpose
     """
 
-    def __init__(self, config, auto_retry=1):
+    def __init__(self, config, auto_retry=1, pool_size=20):
         # .env fed config from rcon.SERVER_INFO
         self.config = config
-        self.conn = None
-        # self._connect()
         self.auto_retry = auto_retry
+        self.mu = threading.RLock()
+        self.idles: list[HLLConnection] = []
+        self.numOpen = 0
+        self.maxOpen = 20
+        self.maxIdle = pool_size
+
+    @contextmanager
+    def with_connection(self) -> HLLConnection:
+        logger.debug("Waiting to acquire lock")
+        with self.mu:
+            if len(self.idles) != 0:
+                logger.debug("acquiring connection from idle pool")
+                conn = self.idles.pop()
+
+            elif self.numOpen >= self.maxOpen:
+                logger.debug("Max connections already open, waiting for connection returned to pool")
+                c = 0
+                while len(self.idles) == 0:
+                    if c >= 30:
+                        logger.error("waiting for connection returned to pool timed out after " + str(c) + " seconds")
```

Comment:
Best to use the string format builtin in the logger.

` logger.error("waiting for connection returned to pool timed out after %s seconds", c)`